src/check/constrain/generate/control_flow.rs

```diff
@@ -50,19 +49,19 @@ pub fn gen_flow(
             let if_expr_exp = Expected::from(ast);
 
             constr.branch_point();
-            generate(then, env, ctx, constr)?;
+            let then_env = generate(then, env, ctx, constr)?;
             if env.is_expr {
                 constr.add("then branch equal to if", &if_expr_exp, &Expected::from(then), env);
             }
 
             constr.branch("if else branch", el.pos);
-            generate(el, env, ctx, constr)?;
+            let else_env = generate(el, env, ctx, constr)?;
             if env.is_expr {
                 constr.add("else branch equal to if", &if_expr_exp, &Expected::from(el), env);
             }
 
             constr.reset_branches();
-            Ok(env.clone())
+            Ok(env.intersection(&then_env.union(&else_env)))
```

Comment:
A comment perhaps to explain why we have an intersection, might not be obvious at first sight.
Or, better still better documentation for both `intersection` and `union` with an example.

Reply:
Actually the documentation for now appears clear.